src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -94,247 +57,133 @@ public override long Position
         => throw new NotSupportedException();
 
     /// <inheritdoc/>
-    protected override void Dispose(bool disposing)
+    public override int ReadByte()
     {
-        if (this.isDisposed)
-        {
-            return;
-        }
-
-        try
-        {
-            this.isDisposed = true;
-            if (disposing)
-            {
-                ReleaseMemoryChunks(this.memoryChunk);
-            }
-
-            this.memoryChunk = null;
-            this.writeChunk = null;
-            this.readChunk = null;
-            this.chunkCount = 0;
-        }
-        finally
+        this.EnsureNotDisposed();
+        if (this.position >= this.length)
         {
-            base.Dispose(disposing);
+            return -1;
         }
-    }
 
-    /// <inheritdoc/>
-    public override void Flush()
-    {
+        _ = this.Read(this.singleByteBuffer, 0, 1);
+        return MemoryMarshal.GetReference<byte>(this.singleByteBuffer);
     }
 
     /// <inheritdoc/>
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public override int Read(byte[] buffer, int offset, int count)
     {
+        this.EnsureNotDisposed();
+
         Guard.NotNull(buffer, nameof(buffer));
         Guard.MustBeGreaterThanOrEqualTo(offset, 0, nameof(offset));
         Guard.MustBeGreaterThanOrEqualTo(count, 0, nameof(count));
 
         const string bufferMessage = "Offset subtracted from the buffer length is less than count.";
         Guard.IsFalse(buffer.Length - offset < count, nameof(buffer), bufferMessage);
 
-        return this.ReadImpl(buffer.AsSpan(offset, count));
+        return this.Read(buffer.AsSpan(offset, count));
     }
 
     /// <inheritdoc/>
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public override int Read(Span<byte> buffer) => this.ReadImpl(buffer);
-
-    private int ReadImpl(Span<byte> buffer)
+    public override int Read(Span<byte> buffer)
     {
         this.EnsureNotDisposed();
 
-        if (this.readChunk is null)
-        {
-            if (this.memoryChunk is null)
-            {
-                return 0;
-            }
+        int offset = 0;
+        int count = buffer.Length;
 
-            this.readChunk = this.memoryChunk;
-            this.readOffset = 0;
+        long remaining = this.length - this.position;
+        if (remaining <= 0)
+        {
+            // Already at the end of the stream, nothing to read
+            return 0;
         }
 
-        IMemoryOwner<byte> chunkBuffer = this.readChunk.Buffer;
-        int chunkSize = this.readChunk.Length;
-        if (this.readChunk.Next is null)
+        if (remaining > count)
         {
-            chunkSize = this.writeOffset;
+            remaining = count;
         }
 
+        int bytesToRead = (int)remaining;
         int bytesRead = 0;
-        int offset = 0;
-        int count = buffer.Length;
-        while (count > 0)
+        while (bytesToRead != 0 && this.currentChunk != this.memoryChunkBuffer.Length)
```

Comment:
I don't know enough about the underlying implementation here; if this is doing additional downstream reads, you might prefer to exit after the first read (you're only required to read "some" data - you don't need to fill the supplied data, just return at least 1 byte or 0 for EOF); if all the data is *already* loaded, I wonder whether your `memoryChunkBuffer` is duplicating the innards of `ReadOnlySequence<byte>` - that already has all the `Slice`, `CopyTo` etc you might want; just a suggestion, though (I can help you grok `ReadOnlySequence<T>` if you're not already familiar)

Reply:
Yeah, looking at `MemoryChunkBuffer` there is definitely some overlap. Ideally, I should be tracking the buffer and chunk indexes internally within that class. 

However, this type needs to be expandable on-demand which AFAIK is not possible with `ReadonlySequence<T>`.

Reply:
The underlying buffer-chain is as mutable as you want it to be (it is your chain, ultimately); if you want to resize it, that is usually as simple as simply creating a new ROS (which is a lightweight struct just tracking the start and end), specifying the new bounds. The chin bits aren't trivial, but not too complex. I guess if what you already have works well, it might be overkill to touch it, though.